File: cli/python/src/memgo_cli/cli/commands/memory/list.py
"""记忆 list 用例。"""

from __future__ import annotations

import time as _time

import typer
from rich.console import Console

from memgo_cli.backend.types import Backend
from memgo_cli.output.branding import (
    print_error,
    print_info,
    timed_status,
)
from memgo_cli.output.format import (
    format_agent_envelope,
    format_memories_table,
    format_memories_text,
    print_result_summary,
)

console = Console()
err_console = Console(stderr=True)


def cmd_list(
    backend: Backend,
    *,
    user_id: str | None,
    agent_id: str | None,
    app_id: str | None,
    run_id: str | None,
    page: int,
    page_size: int,
    category: str | None,
    after: str | None,
    before: str | None,
    show_expired: bool,
    latest_only: bool,
    output: str,
) -> None:
    """列出指定范围的记忆。"""
    from memgo_cli.runtime.state import is_agent_mode, set_current_command

    set_current_command("list")
    if is_agent_mode():
        output = "agent"
    if page_size < 1:
        print_error(err_console, "--page-size must be >= 1.", hint=None)
        raise typer.Exit(1)
    if page < 1:
        print_error(err_console, "--page must be >= 1.", hint=None)
        raise typer.Exit(1)

    _start = _time.perf_counter()
    with timed_status(err_console, "Listing memories...") as _ts:
        try:
            results = backend.list_memories(
                user_id=user_id,
                agent_id=agent_id,
                app_id=app_id,
                run_id=run_id,
                page=page,
                page_size=page_size,
                category=category,
                after=after,
                before=before,
                show_expired=show_expired,
                latest_only=latest_only,
            )
        except Exception as e:
            print_error(err_console, str(e), hint=None)
            raise typer.Exit(1) from None
    _elapsed = _time.perf_counter() - _start

    if output == "quiet":
        return

    if output in ("json", "agent"):
        scope = {
            k: v
            for k, v in {
                "user_id": user_id,
                "agent_id": agent_id,
                "app_id": app_id,
                "run_id": run_id,
            }.items()
            if v
        }
        format_agent_envelope(
            console,
            command="list",
            data=results,
            scope=scope or None,
            count=len(results),
            duration_ms=int(_elapsed * 1000),
        )
    elif output == "table":
        if results:
            format_memories_table(console, results, show_score=False)
            print_result_summary(
                console,
                len(results),
                duration_secs=_elapsed,
                page=page,
                user_id=user_id,
                agent_id=agent_id,
            )
        else:
            console.print()
            print_info(console, "No memories found.")
            console.print()
    else:
        if results:
            format_memories_text(console, results, title="memories")
            print_result_summary(
                console,
                len(results),
                duration_secs=_elapsed,
                page=page,
                user_id=user_id,
                agent_id=agent_id,
            )
        else:
            console.print()
            print_info(console, "No memories found.")
            console.print()

File: cli/node/src/cli/commands/memory/list.ts
import type { JsonObject } from '../../../backend/json.js';
import type { Backend } from '../../../backend/types.js';
import { printError, printInfo, timedStatus } from '../../../output/branding.js';
import {
  formatAgentEnvelope,
  formatMemoriesTable,
  formatMemoriesText,
  printResultSummary,
} from '../../../output/format.js';
import { exit } from '../../../runtime/exit.js';
import { setCurrentCommand } from '../../../runtime/state.js';

/**
 * 按实体范围与分页条件列出记忆。
 * @param backend - 已装配的后端连接器。
 * @param opts - 本次操作选项。
 * @param opts.userId - 用户 ID。
 * @param opts.agentId - 代理 ID。
 * @param opts.appId - 应用 ID。
 * @param opts.runId - 会话 ID。
 * @param opts.page - 分页页码。
 * @param opts.pageSize - 每页条目数。
 * @param opts.category - 分类筛选条件。
 * @param opts.after - 创建日期下界。
 * @param opts.before - 创建日期上界。
 * @param opts.showExpired - 包含已过期记忆的选项。
 * @param opts.latestOnly - 只返回最新版本的选项。
 * @param opts.output - 终端输出格式。
 */
export async function cmdList(
  backend: Backend,
  opts: {
    userId?: string;
    agentId?: string;
    appId?: string;
    runId?: string;
    page: number;
    pageSize: number;
    category?: string;
    after?: string;
    before?: string;
    showExpired?: boolean;
    latestOnly?: boolean;
    output: string;
  }
): Promise<void> {
  setCurrentCommand('list');
  if (opts.pageSize < 1) {
    printError('--page-size must be >= 1.');
    exit(1);
  }
  if (opts.page < 1) {
    printError('--page must be >= 1.');
    exit(1);
  }

  const start = performance.now();
  let results: JsonObject[];
  try {
    results = await timedStatus('Listing memories...', async () => {
      return backend.listMemories({
        userId: opts.userId,
        agentId: opts.agentId,
        appId: opts.appId,
        runId: opts.runId,
        page: opts.page,
        pageSize: opts.pageSize,
        category: opts.category,
        after: opts.after,
        before: opts.before,
        showExpired: opts.showExpired,
        latestOnly: opts.latestOnly,
      });
    });
  } catch (e) {
    printError(e instanceof Error ? e.message : String(e));
    exit(1);
  }
  const elapsed = (performance.now() - start) / 1000;

  if (opts.output === 'quiet') return;

  if (opts.output === 'agent' || opts.output === 'json') {
    const scope: Record<string, string | undefined> = {
      user_id: opts.userId,
      agent_id: opts.agentId,
      app_id: opts.appId,
      run_id: opts.runId,
    };
    formatAgentEnvelope({
      command: 'list',
      data: results,
      scope,
      count: results.length,
      durationMs: Math.round(elapsed * 1000),
    });
  } else if (opts.output === 'table') {
    if (results.length > 0) {
      formatMemoriesTable(results, {});
      printResultSummary({
        count: results.length,
        durationSecs: elapsed,
        page: opts.page,
        scopeIds: { user_id: opts.userId, agent_id: opts.agentId },
      });
    } else {
      console.log();
      printInfo('No memories found.');
      console.log();
    }
  } else {
    if (results.length > 0) {
      formatMemoriesText(results, 'memories');
      printResultSummary({
        count: results.length,
        durationSecs: elapsed,
        page: opts.page,
        scopeIds: { user_id: opts.userId, agent_id: opts.agentId },
      });
    } else {
      console.log();
      printInfo('No memories found.');
      console.log();
    }
  }
}
